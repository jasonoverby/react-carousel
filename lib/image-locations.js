export const imageLocations = [
  {
    src: 'https://bit.ly/370FSY3',
    href:
      'https://www.scientificamerican.com/article/cats-can-get-coronavirus-study-suggests-but-pet-owners-need-not-panic/',
    alt: 'cat on a bed',
  },
  {
    src: 'https://bit.ly/2Z2kVYj',
    href:
      'https://www.humanesociety.org/resources/understanding-feline-language',
    alt: 'playful cat',
  },
  {
    src: 'https://bit.ly/3p9Bko8',
    href: 'https://pets.webmd.com/cats/ss/slideshow-is-my-cat-normal',
    alt: 'staring cat',
  },
];
