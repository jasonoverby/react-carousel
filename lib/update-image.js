import { imageLocations } from './image-locations';
import { wait } from './wait';

const NUM_IMAGES = imageLocations.length;

export const getNewImageIndex = (imageIndex, incrementOrDecrement) => {
  if (incrementOrDecrement === 'increment') {
    return imageIndex === NUM_IMAGES - 1 ? 0 : imageIndex + 1;
  }
  if (incrementOrDecrement === 'decrement') {
    return imageIndex === 0 ? NUM_IMAGES - 1 : imageIndex - 1;
  }

  return imageIndex;
};

export const updateImage = async ({
  updatedImageIndex,
  setImageIndex,
  setImageLoaded,
}) => {
  // hide current image
  setImageLoaded(false);
  // gives the old image time to ease out
  await wait(200);
  setImageIndex(updatedImageIndex);
  // gives the new image time to ease in
  await wait(500);
  // show updated image
  setImageLoaded(true);
};
