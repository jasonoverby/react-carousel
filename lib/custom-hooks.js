import { useEffect } from 'react';

import { getNewImageIndex, updateImage } from './utils';

export const useImageCycling = ({
  carouselPaused,
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => {
  useEffect(() => {
    const intervalId = setInterval(async () => {
      if (!carouselPaused) {
        const updatedImageIndex = getNewImageIndex(imageIndex, 'increment');
        await updateImage({
          setActiveDotIndex,
          setImageIndex,
          setImageLoaded,
          updatedImageIndex,
        });
      }
    }, 10000);
    return () => clearInterval(intervalId);
  });
};
