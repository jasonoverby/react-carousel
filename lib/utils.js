import { imageLocations } from '/lib/image-locations';

const NUM_IMAGES = imageLocations.length;

export const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

export const getNewImageIndex = (imageIndex, incrementOrDecrement) => {
  if (incrementOrDecrement === 'increment') {
    return imageIndex === NUM_IMAGES - 1 ? 0 : imageIndex + 1;
  } else if (incrementOrDecrement === 'decrement') {
    return imageIndex === 0 ? NUM_IMAGES - 1 : imageIndex - 1;
  }
};

export const updateImage = async ({
  updatedImageIndex,
  setImageIndex,
  setImageLoaded,
}) => {
  setImageLoaded(false);
  await wait(500);
  setImageIndex(updatedImageIndex);
  await wait(200);
  setImageLoaded(true);
};
