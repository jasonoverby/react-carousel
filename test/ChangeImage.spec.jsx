import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import * as utils from '../lib/utils';
import {
  DecrementImage,
  IncrementImage,
  UpdateImage,
} from '../components/ChangeImage';

jest.spyOn(utils, 'updateImage');
jest.spyOn(utils, 'getNewImageIndex');

describe('ChangeImage', () => {
  const setActiveDotIndex = jest.fn();
  const setImageIndex = jest
    .fn()
    .mockImplementation((imageIndex, incrementOrDecrement) =>
      incrementOrDecrement === 'increment' ? imageIndex + 1 : imageIndex - 1,
    );
  const setImageLoaded = jest.fn();

  describe('UpdateImage', () => {
    const getContainer = (incrementOrDecrement, imageIndex) =>
      render(
        <UpdateImage
          incrementOrDecrement={incrementOrDecrement}
          imageIndex={imageIndex}
          setActiveDotIndex={setActiveDotIndex}
          setImageIndex={setImageIndex}
          setImageLoaded={setImageLoaded}
        />,
      );

    it('increments the image', () => {
      const imageIndex = 0;
      const incrementOrDecrement = 'increment';
      const updatedImageIndex = imageIndex + 1;
      const { container } = getContainer(incrementOrDecrement, imageIndex);

      const button = container.firstChild;
      expect(button.getAttribute('aria-label')).toEqual('show next image');
      expect(button.firstChild.textContent).toEqual('▶');
      fireEvent.click(button);
      expect(utils.getNewImageIndex).toHaveBeenCalledWith(
        imageIndex,
        incrementOrDecrement,
      );
      expect(utils.updateImage).toHaveBeenCalledWith({
        setActiveDotIndex,
        setImageIndex,
        setImageLoaded,
        updatedImageIndex,
      });
    });

    it('decrements the image', () => {
      const imageIndex = 1;
      const incrementOrDecrement = 'decrement';
      const updatedImageIndex = imageIndex - 1;
      const { container } = getContainer(incrementOrDecrement, imageIndex);

      const button = container.firstChild;
      expect(button.getAttribute('aria-label')).toEqual('show previous image');
      expect(button.firstChild.textContent).toEqual('◀');
      fireEvent.click(button);
      expect(utils.getNewImageIndex).toHaveBeenCalledWith(
        imageIndex,
        incrementOrDecrement,
      );
      expect(utils.updateImage).toHaveBeenCalledWith({
        setActiveDotIndex,
        setImageIndex,
        setImageLoaded,
        updatedImageIndex,
      });
    });
  });

  describe('DecrementImage', () => {
    const imageIndex = 1;
    it('renders correctly', () => {
      const { asFragment } = render(
        <DecrementImage
          imageIndex={imageIndex}
          setActiveDotIndex={setActiveDotIndex}
          setImageIndex={setImageIndex}
          setImageLoaded={setImageLoaded}
        />,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('IncrementImage', () => {
    const imageIndex = 1;
    it('renders correctly', () => {
      const { asFragment } = render(
        <IncrementImage
          imageIndex={imageIndex}
          setActiveDotIndex={setActiveDotIndex}
          setImageIndex={setImageIndex}
          setImageLoaded={setImageLoaded}
        />,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
