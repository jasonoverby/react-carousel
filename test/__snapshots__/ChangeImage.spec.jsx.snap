// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeImage DecrementImage renders correctly 1`] = `
<DocumentFragment>
  <button
    aria-label="show previous image"
    class="changeImage"
    type="button"
  >
    <span
      class="arrow"
    >
      ◀
    </span>
  </button>
</DocumentFragment>
`;

exports[`ChangeImage IncrementImage renders correctly 1`] = `
<DocumentFragment>
  <button
    aria-label="show next image"
    class="changeImage"
    type="button"
  >
    <span
      class="arrow"
    >
      ▶
    </span>
  </button>
</DocumentFragment>
`;
